lazy-pages/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+        (wasm_page, wasm_page_native_addr)
+    };
+
+    let page_info = LAZY_PAGES_INFO.with(|info| info.borrow_mut().remove(&wasm_page.raw()));
+    if page_info.is_none() {
+        // TODO: we need to do something here. May be throw it to old sig handler.
+        panic!("sigsegv/sigbus from unknown memory");
+    }
+
+    let res = unsafe {
+        libc::mprotect(
+            wasm_page_native_addr as *mut libc::c_void,
+            PAGE_SIZE,
+            libc::PROT_READ | libc::PROT_WRITE,
+        )
+    };
+    assert!(
+        res == 0,
+        "Cannot remove page protection, unexpected os behavior: {}",
+        errno::errno()
+    );
+
+    let page_as_slice =
+        unsafe { std::slice::from_raw_parts_mut(wasm_page_native_addr as *mut u8, PAGE_SIZE) };
+    let hash_key_in_storage = page_info.unwrap();
+    let res = sp_io::storage::read(&hash_key_in_storage, page_as_slice, 0);
+    assert!(res.is_some(), "Wasm page must have data in storage");
+    assert!(
+        res.unwrap() as usize == PAGE_SIZE,
+        "Page data must contain {} bytes, actually has {}",
+        PAGE_SIZE,
+        res.unwrap()
+    );
+
+    let res = RELEASED_LAZY_PAGES.with(|rpages| {
+        rpages
+            .borrow_mut()
+            .insert(wasm_page.raw(), page_as_slice.to_vec())
+    });
+    assert!(res.is_none(), "Any page cannot be released twice");
+
+    log::debug!(target: "gear_node::sig_handler", "Finish signal handling");
+}
+
+/// Save page key in storage
+pub fn save_page_lazy_info(wasm_page: u32, key: &[u8]) {
+    LAZY_PAGES_INFO
+        .with(|lazy_pages_info| lazy_pages_info.borrow_mut().insert(wasm_page, key.to_vec()));
+}
+
+/// Returns vec of not-accessed wasm lazy pages
+pub fn get_wasm_lazy_pages_numbers() -> sp_std::vec::Vec<u32> {
+    LAZY_PAGES_INFO.with(|lazy_pages_info| lazy_pages_info.borrow().iter().map(|x| *x.0).collect())
+}
+
+/// Set current wasm memory begin addr
+pub fn set_wasm_mem_begin_addr(wasm_mem_begin: usize) {
+    WASM_MEM_BEGIN.with(|x| *x.borrow_mut() = wasm_mem_begin);
+}
+
+/// Reset lazy pages info
+pub fn reset_lazy_pages_info() {
+    LAZY_PAGES_INFO.with(|x| x.replace(BTreeMap::new()));
+    RELEASED_LAZY_PAGES.with(|x| x.replace(BTreeMap::new()));
+    WASM_MEM_BEGIN.with(|x| x.replace(0));
+}
+
+/// Returns vec of lazy pages which has been accessed
+pub fn get_released_pages() -> sp_std::vec::Vec<u32> {
+    RELEASED_LAZY_PAGES.with(|x| x.borrow().iter().map(|x| *x.0).collect())
+}
+
+/// Returns page data which page has in storage before execution
+pub fn get_released_page_old_data(page: u32) -> sp_std::vec::Vec<u8> {
+    RELEASED_LAZY_PAGES.with(|x| x.borrow().get(&page).expect("Must have this page").to_vec())
+}
+
+/// Initialize lazy pages:
+/// 1) checks whether lazy pages is supported in current environment
+/// 2) set signals handler
+///
+/// # Safety
+/// use OS specific functions
+pub unsafe fn init_lazy_pages() -> bool {
```

Comment:
Maybe it is better ro return `Result` with corresponding errors?

Reply:
Yes, good idea, will do this today

Reply:
A, no. When start to this - release that actually it's unused. Cannot find, a point at which runtime should now the problem in native part. Information whether we can or not enable lazy pages must be enough